src/Type/ArrayType.php

```diff
@@ -286,9 +286,11 @@ public function setOffsetValueType(?Type $offsetType, Type $valueType, bool $uni
 
 		$array = new self(
 			TypeCombinator::union($this->keyType, $offsetType),
-			$unionValues ? TypeCombinator::union($this->itemType, $valueType) : $valueType,
+			$unionValues || $offsetType instanceof ConstantStringType || $offsetType instanceof ConstantIntegerType
+				? TypeCombinator::union($this->itemType, $valueType)
+				: $valueType,
 		);
-		if ($offsetType instanceof ConstantIntegerType || $offsetType instanceof ConstantStringType) {
+		if ($offsetType instanceof ConstantStringType || $offsetType instanceof ConstantIntegerType) {
```

Comment:
Why did you switch around the order of these conditions? Does it matter?

Reply:
The order doesn't matter. It was just in this order line 278 and inversed line 291.